lib/async.array.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+const { promisify } = require('util');
+
+const timeout = promisify(resolve => setTimeout(resolve, 0));
+
+class AsyncArray extends Array {
+  constructor(...args) {
+    super(...args);
+
+    this.min = 5;
+    this.ratio = 1.5;
+  }
+
+  set percent(value) {
+    if (value < 1 && value >= 0) {
+      this.ratio = value / (1 - value);
+    } else {
+      throw new Error('The percen must be less than 1 and greater ' +
+        'than or equal to 0');
+    }
+  }
+
+  [Symbol.asyncIterator]() {
+    let sum = 0;
+    let count = 0;
+    let begin = Date.now();
+    let iterMax = this.min;
+
+    const getItem = () => {
+      const done = count >= this.length;
+      const value = done ? undefined : this[count];
```

Comment:
You don't need this check as it is already performed by an array, though no strong feelings about this one.